micro_g/scripts/pose_servoing_controller.py

```diff
@@ -0,0 +1,194 @@
+from interbotix_xs_modules.xs_robot.arm import InterbotixManipulatorXS
+from rclpy.utilities import remove_ros_args
+from tf2_geometry_msgs import do_transform_pose
+
+
+class XSArmPoseServoingController(InterbotixManipulatorXS):
+    """
+    Tracks a published end effector pose.
+    """
+
+    def __init__(
+        self,
+        robot_model: str,
+        robot_name: str,
+        control_update_rate: float = 10.0,
+        moving_time: float = 0.5,
+        xs_args=None,
+    ):
+        """
+        Create the controller
+
+        Args:
+            robot_model: the robot model
+            robot_name: the robot name
+            control_update_rate: the rate at which the controller is updated (Hz).
+            moving_time: the duration in which all movements should take place (s).
+            xs_args: Extra arguments to be passed to InterbotixManipulatorXS. Defaults
+                to None.
+        """
+        # Initialize the interface to the arm
+        InterbotixManipulatorXS.__init__(
+            self,
+            robot_model=robot_model,
+            robot_name=robot_name,
+            start_on_init=True,
+            moving_time=moving_time,
+            accel_time=moving_time / 2.0,
+            args=xs_args,
+            node_name="pose_servoing",
+        )
+
+        # Set the controller update rate
+        self.rate = self.core.create_rate(control_update_rate)
+
+        # Define the transforms we'll use. Notation is T_{to}{from}.
+        # Frames are:
+        #   - s = space frame, fixed to the robot's base link
+        #   - y = yaw frame, same as s but rotated about z by the current waist angle
+        #   - b = body frame, fixed to the end effector
+        #   - d = desired frame for the end effector
+        self.T_sy = np.identity(4)
+        self.T_sd = self.arm.get_ee_pose_command()
+        self.update_frames()
+
+        # Initialize the tf2 buffer and listener
+        self.tf_buffer = tf2_ros.Buffer()
+        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self.core)
+        # Set the target frame for transformation
+        self.target_frame = f"{robot_name}/base_link"
+
+        # Create a subscription to the pose that we will track
+        self.core.create_subscription(
+            PoseStamped, "desired_eef_pose", self.desired_pose_callback, 10
+        )
+        time.sleep(0.5)
+        self.core.get_logger().info("Ready to receive servoing commands.")
+
+    def update_frames(self):
+        """
+        Update the coordinate frames used by this controller.
+
+        We need to do this manually because the y frame (base frame rotated to match
+        arm yaw) is not actually published as a TF frame.
+        """
+        T_sb = self.arm.get_ee_pose_command()  # Current EEF pose in space frame
+        rpy = ang.rotation_matrix_to_euler_angles(T_sb[:3, :3])
+        self.T_sy[:2, :2] = ang.yaw_to_rotation_matrix(rpy[2])
```

Comment:
Should we publish this TF?

Reply:
I'm not sure we need to, since it's currently only used internally by this node.